### DISCUSSION TITLE
Only one token created if `claim_token` called several times simultaneously

### DISCUSSION THREAD

**Discussion by sxurik** [epam/mintgate#74] [General]
Claiming several tokens like this:
```js
await Promise.all([
  bob.contract.claim_token({ gate_id: someGateId }),
  bob.contract.claim_token({ gate_id: someGateId }),
]);
```
should mint 2 tokens.

But for some reason only one token gets minted and added, all calls to `claim_token` return the same `token_id`. 
The same happens when making for example 5 calls instead of 2 - only 1 token is minted.

Making 4 calls from 2 different accounts:
```js
await Promise.all([
  bob.contract.claim_token({ gate_id: someGateId }),
  alice.contract.claim_token({ gate_id: someGateId }),
  bob.contract.claim_token({ gate_id: someGateId }),
  alice.contract.claim_token({ gate_id: someGateId }),
]);
```
results in 2 tokens being created instead of 4.

But if there were some method-calls on the contract recently, `claim_token` from this contract works correctly.

So this code:
```js
await alice.contract.create_collectible({ ... });

await Promise.all([
  bob.contract.claim_token({ gate_id: someGateId }),
  alice.contract.claim_token({ gate_id: someGateId }),
  bob.contract.claim_token({ gate_id: someGateId }),
  alice.contract.claim_token({ gate_id: someGateId }),
]);
```
mints 2 tokens for `alice` (which is expected) and 1 token for `bob` (instead of 2).

**Comment by zahhar** [comment]
However, this results in 2 tokens claimed:

```
await Promise.all([
  bob.contract.claim_token({ gate_id: someGateId })
]);

await Promise.all([
  bob.contract.claim_token({ gate_id: someGateId })
]);
```

**Comment by sxurik** [comment]
I've contacted guys from NEAR on discord ([my comment there](https://discord.com/channels/490367152054992913/817102046725406780/832173471554600990)). 

This is the response I got:
![image](https://user-images.githubusercontent.com/32016927/114992975-00317b00-9ea4-11eb-8e30-c70b3e221375.png)

It doesn't seem to be the problem on our side. So, I guess we are waiting for them to investigate the issue on their end further.

I'll stay in contact with them and post any updates here.

**Comment by sxurik** [comment]
The problem seems to be with `near-api-js`. 
The devs created an [issue](https://github.com/near/near-api-js/issues/559), will begin implementing a fix soon.

**Reply by acuarica** [reply]
Good to hear and great finding @sxurik. I would point out that this only happens the first time. You mentioned it in the discord channel, but the issue might not describe the whole picture. This can help to pinpoint the bug. Here is a small test.

```js
const os = require("os");
const nearApi = require("near-api-js");

run();

async function run() {
    const near = await nearApi.connect({
        keyStore: new nearApi.keyStores.UnencryptedFileSystemKeyStore(
            `${os.homedir()}/.near-credentials`
        ),
        networkId: "default",
        nodeUrl: "https://rpc.testnet.near.org",
        walletUrl: "https://wallet.testnet.near.org",
        helperUrl: "https://helper.testnet.near.org",
    });

    const account = await near.account("alice-1618489554879-389812");

    const contract = new nearApi.Contract(account, "nft-1618489534408-352163", {
        changeMethods: ["claim_token"],
    });

    const args = { gate_id: 'A5L5MWCCTBI5' };

    const ids = await Promise.all([contract.claim_token(args), contract.claim_token(args)]);
    console.log(ids);

    const ids2 = await Promise.all([contract.claim_token(args), contract.claim_token(args)]);
    console.log(ids2);
}
```

Giving the following output:

```sh
Receipt: 95Zti8PH6vo6VptAZY7FXfTp5NNAxL87Kya4qVBrKCNK
        Log [nft-1618489534408-352163]: claim_token
Receipt: 95Zti8PH6vo6VptAZY7FXfTp5NNAxL87Kya4qVBrKCNK
        Log [nft-1618489534408-352163]: claim_token
[ '28', '28' ]
Receipt: CbUYN9efoWk1doyoPFiPQpsAGNZeivgBXgV2vHx9YyP9
        Log [nft-1618489534408-352163]: claim_token
Receipt: 5QT14mmzGZAKX9C5VAiJp4XoYGSfJ2hbg1R3xGEhY7qT
        Log [nft-1618489534408-352163]: claim_token
[ '29', '30' ]
```

The first time, we got `28` twice as the issue mentions. However the second time we got the expected different values.
